{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Cohort: # 04\n",
    "\n",
    "# Team No.: # 10\n",
    "\n",
    "## Members:\n",
    "* Loh De Rong (1003557)\n",
    "* Hadi Alwi (1003482)\n",
    "* Ishika Kajaria (1003555)\n",
    "* Pham Trung Viet (1003589)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "import math"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part 1 - Getting the Cost Function Ready"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 1\n",
    "\n",
    "Instance variables are being instantiated. The init method initializes all parameters to zero. All penalty parameters are set to a default value of 10."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "class PrimerDesign(object):\n",
    "    \n",
    "    def __init__ (self, name):\n",
    "        \n",
    "        '''parameters for the length criterion'''\n",
    "        self.max_length = 0\n",
    "        self.min_length = 0\n",
    "        self.penalty_length = 10\n",
    "        \n",
    "        '''parameters for the temperature difference criterion'''\n",
    "        self.max_tdiff = 0\n",
    "        self.min_tdiff = 0\n",
    "        self.penalty_tdiff = 10\n",
    "        \n",
    "        '''parameters for the cg content criterion'''\n",
    "        self.max_cg = 0\n",
    "        self.min_cg = 0\n",
    "        self.penalty_cg = 10\n",
    "        \n",
    "        '''parameters for the annealing temperature criterion'''\n",
    "        self.max_temp = 0\n",
    "        self.min_temp = 0\n",
    "        self.penalty_temp = 10\n",
    "        \n",
    "        '''parameters for the run criterion'''\n",
    "        self.run_threshold = 0\n",
    "        self.penalty_runs = 10\n",
    "        \n",
    "        '''parameters for the repeat criterion'''\n",
    "        self.repeat_threshold = 0\n",
    "        self.penalty_repeats = 10\n",
    "        \n",
    "        '''parameters for the specificity criterion'''\n",
    "        self.penalty_specificity = 10 \n",
    "        \n",
    "        '''locations where the forward primer should be chosen from'''\n",
    "        self.fp_start = 0\n",
    "        self.fp_end = 0\n",
    "        \n",
    "        '''locations where the reverse primer should be chosen from'''\n",
    "        self.rp_start = 0\n",
    "        self.rp_end = 0\n",
    "        \n",
    "        ''' parameters for the simulated annealing portion'''\n",
    "        self.initial_temperature = 200\n",
    "        self.stopping_temperature = 0.01\n",
    "        self.drop_fraction = 0.999\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 2\n",
    "\n",
    "The set_dna_sequence method takes in a raw gene sequence, removes the numbers and spaces, and returns only a, t, c, g in lower case characters. The gene will be considered as invalid if it is not in a string format, contains upper case characters, or characters that are not a, t, c, g."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "class PrimerDesign(PrimerDesign): \n",
    "    \n",
    "    def set_dna_sequence(self, dna_sequence):\n",
    "        \n",
    "        # check if dna_sequence is a string\n",
    "        if not isinstance(dna_sequence, str):\n",
    "            return \"ERROR\"\n",
    "        \n",
    "        dna_sequence = dna_sequence.replace(\" \", \"\") # removing white spaces in the string\n",
    "        dna_sequence = dna_sequence.replace(\"\\n\", \"\") # removing new lines in the string\n",
    "        edited_dna_sequence = \"\"\n",
    "        \n",
    "        for base in dna_sequence:\n",
    "            if base in \"atcg\":\n",
    "                edited_dna_sequence += base # adding base\n",
    "            elif base not in ['0','1','2','3','4','5','6','7','8','9']:\n",
    "                return \"ERROR\" # check that base is not any other alphabet besides a,t,c,g\n",
    "        \n",
    "        # initializes dna_sequence if all conditions met\n",
    "        self.dna_sequence = edited_dna_sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Testing method in Task 2'''\n",
    "\n",
    "# Test Case 1: check that the dna_sequence containing only a,t,c,g in lower case is instantiated correctly\n",
    "DNA2_1 = PrimerDesign(\"DNA2_1\")\n",
    "DNA2_1.set_dna_sequence(\"atcgcgatcg\")\n",
    "assert DNA2_1.dna_sequence == \"atcgcgatcg\"\n",
    "\n",
    "# Test Case 2: check that the dna_sequence is stripped off its numbers, newlines and spaces, and instantiated correctly\n",
    "DNA2_2 = PrimerDesign(\"DNA2_2\")\n",
    "DNA2_2.set_dna_sequence(\"1 atc\\n 2atac\\ngt  3tag 4ccgg \")\n",
    "assert DNA2_2.dna_sequence == \"atcatacgttagccgg\"\n",
    "\n",
    "# Test Case 3: check that the dna_sequence is of type str\n",
    "DNA2_3 = PrimerDesign(\"DNA2_3\")\n",
    "assert DNA2_3.set_dna_sequence([\"this does not work\"]) == \"ERROR\"\n",
    "\n",
    "# Test Case 4: check that the dna_sequence only contains a,t,c,g in lower case\n",
    "DNA2_4 = PrimerDesign(\"DNA2_4\")\n",
    "assert DNA2_4.set_dna_sequence(\"aTcggggatC\") == \"ERROR\"\n",
    "\n",
    "# Test Case 5: check that the dna_sequence does not contain other alphabets besides a,t,c,g\n",
    "DNA2_5 = PrimerDesign(\"DNA2_4\")\n",
    "assert DNA2_5.set_dna_sequence(\"kcggtfaggat\") == \"ERROR\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 3\n",
    "\n",
    "The method func_select_random aims to select a forward primer or reverse primer at random with a specified length."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "class PrimerDesign(PrimerDesign):\n",
    "    \n",
    "    def func_select_random(self, sqtype='forward', length = 20):\n",
    "        \n",
    "        # primer should be either forward or reverse\n",
    "        # the length has to be a positive number\n",
    "        if (sqtype != 'forward' and sqtype != 'reverse') or (length < 0 or length > len(self.dna_sequence)):\n",
    "            return \"ERROR\"\n",
    "        \n",
    "        # returns empty string if length = 0\n",
    "        if length == 0:\n",
    "            return \"\"\n",
    "        \n",
    "        elif(sqtype == 'forward'):\n",
    "            start_limit = self.fp_start \n",
    "            end_limit = self.fp_end \n",
    "            \n",
    "        elif(sqtype == 'reverse'):\n",
    "        \n",
    "            start_limit = self.rp_start \n",
    "            end_limit = self.rp_end\n",
    "            \n",
    "        # generate a random starting position that lies between these limits\n",
    "        starting_point = random.randrange(start_limit, end_limit - length + 1)\n",
    "        \n",
    "        # adding bases to obtain the full-length primer\n",
    "        primer = \"\"\n",
    "        for i in range(length):\n",
    "            primer += self.dna_sequence[starting_point + i]\n",
    "        \n",
    "        return primer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Testing method in Task 3'''\n",
    "\n",
    "DNA3 = PrimerDesign(\"DNA3\")\n",
    "DNA3.set_dna_sequence(\"aaaaaaaaaattttttttttccccccccccggggggggggaaaaaaaaaattttttttttccccccccccgggggggggg\")\n",
    "DNA3.fp_start = 5\n",
    "DNA3.fp_end = 35\n",
    "DNA3.rp_start = 50\n",
    "DNA3.rp_end = 80\n",
    "\n",
    "# Test Case 1: check that sqtype is either forward or reverse\n",
    "assert DNA3.func_select_random(sqtype = \"thisdoesnotwork\") == \"ERROR\"\n",
    "\n",
    "# Test Case 2: check that length > 0\n",
    "assert DNA3.func_select_random(length = -5) == \"ERROR\"\n",
    "\n",
    "# Test Case 3: check that length <= dna_sequence\n",
    "assert DNA3.func_select_random(length = 2000) == \"ERROR\"\n",
    "\n",
    "# Test Case 4: check that a random forward primer sequence is selected within limits\n",
    "assert DNA3.func_select_random(sqtype='forward', length = 20) in \"aaaaattttttttttccccccccccggggg\"\n",
    "\n",
    "# Test Case 5: check that a random reverse primer sequence is selected within limits\n",
    "assert DNA3.func_select_random(sqtype='reverse', length = 20) in \"ttttttttttccccccccccgggggggggg\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 4\n",
    "\n",
    "The following methods calculate the properties for a given primer string. We assume that the primer sequence is valid because it will be called using the func_select_random method. Also note that there is no need to reverse complement to find the reverse primer because the cost functions will effectively give the same value e.g. the cg fraction will be the same whether you reverse complement it or not. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "class PrimerDesign(PrimerDesign): \n",
    "    \n",
    "    # calculates the functional length of the primer sequence\n",
    "    def func_length(self, sq):\n",
    "        return len(sq)\n",
    "    \n",
    "    # calculates the fraction of cg bases in the primer sequence\n",
    "    def func_cg_fraction(self, sq):\n",
    "        if len(sq) == 0:\n",
    "            return 0\n",
    "        return (sq.count(\"c\") + sq.count(\"g\"))/len(sq)\n",
    "    \n",
    "    # calculates the annealing temperature of the primer\n",
    "    def func_temperature(self,sq):\n",
    "        return 4*(sq.count(\"c\") + sq.count(\"g\")) + 2*(sq.count(\"a\") + sq.count(\"t\"))           "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "class PrimerDesign(PrimerDesign):\n",
    "\n",
    "    # calculates the number of base runs that exceed the acceptable threshold\n",
    "    def func_count_runs(self,sq):\n",
    "        R_max = self.run_threshold\n",
    "        n_runs = 0\n",
    "        count = 1\n",
    "        \n",
    "        for i in range(len(sq)-1):\n",
    "            # adds 1 to count if the next element is the same as the previous\n",
    "            if sq[i] == sq[i+1]:\n",
    "                count += 1\n",
    "            else:\n",
    "                # adds 1 to base run if the cumulative count exceeds threshold\n",
    "                # then reset the counter for the next base\n",
    "                if count > R_max:\n",
    "                    n_runs += 1\n",
    "                count = 1 # counter reset for the next base\n",
    "        return n_runs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "class PrimerDesign(PrimerDesign):\n",
    "    \n",
    "    # calculates the number of repeats detected\n",
    "    # repeats refer to repeating sequences of two bases in the primer\n",
    "    def func_count_repeats(self,sq):\n",
    "        \n",
    "        di_repeats = ['at','ac','ag','ca','ct','cg','ga','gt','gc','ta','tc','tg']\n",
    "        count = 0\n",
    "        \n",
    "        for i in range(0, len(sq)-2, 2):\n",
    "            # adds 1 to count if the next sequence is a repeat of the previous\n",
    "            # starting from the 0th element\n",
    "            current_base1 = sq[i:i+2]\n",
    "            next_base1 = sq[i+2:i+4]\n",
    "            if next_base1 == current_base1 and current_base1 in di_repeats:\n",
    "                count += 1\n",
    "            \n",
    "            # adds 1 to count if the next sequence is a repeat of the previous\n",
    "            # starting from the 1st element\n",
    "            current_base2 = sq[i+1:i+3]\n",
    "            next_base2 = sq[i+3:i+5]\n",
    "            if next_base2 == current_base2 and current_base2 in di_repeats:\n",
    "                count += 1    \n",
    "        return count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "class PrimerDesign(PrimerDesign):\n",
    "    def func_count_specificity(self,sq):\n",
    "        n_positions = 0\n",
    "       \n",
    "        #iterate through dna_sequence and count the number of primer occurences\n",
    "        for i in range(len(self.dna_sequence) - len(sq) +1):\n",
    "            if sq == self.dna_sequence[i:i+len(sq)]:\n",
    "                n_positions += 1\n",
    "        return n_positions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Test Cases for 15 test cases are provided here.\n",
    "\n",
    "## (from eDimension)\n",
    "### case 1. 'cattaaaaatacgaaaaaagtcat'\n",
    "It has length 24, cg fraction of 5/24, temperature of 4(5) + 2(19) = 58, two runs of a no repeats, hence a repeat score of 0.\n",
    "\n",
    "### case 2. 'atatatatattttatatataa'\n",
    "It has length 21, cg fraction of 0/21, temperature of 4(0) + 2(21) = 42, no runs of a or t. For repeats, five instances of at, four instances of ta, followed by four instances of ta and three instances of at. Thus the repeat score is 12.\n",
    "\n",
    "### case 3 - edge case '' (empty string).\n",
    "It has length 0, cg fraction 0, temperature 0, no runs, no repeats.\n",
    "\n",
    "## (own test case)\n",
    "### case 4. 'aattttggccccctgact'\n",
    "It has length 20, cg fraction 10/20, temperature of 4(10) + 2(10) = 60, 1 run of c. For repeats, three instances of ga and two instances of ag. Thus the repeat score is 3."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Testing methods in Task 4'''\n",
    "\n",
    "DNA4 = PrimerDesign(\"DNA4\")\n",
    "DNA4.run_threshold = 4\n",
    "sq1 = 'cattaaaaatacgaaaaaagtcat'\n",
    "sq2 = 'atatatatattttatatataa'\n",
    "sq3 = ''\n",
    "sq4 = 'aattttggccccctgagaga'\n",
    "\n",
    "# Test Cases 1-4: check length\n",
    "assert DNA4.func_length(sq1) == 24\n",
    "assert DNA4.func_length(sq2) == 21\n",
    "assert DNA4.func_length(sq3) == 0\n",
    "assert DNA4.func_length(sq4) == 20\n",
    "\n",
    "# Test Cases 5-8: check cg fraction\n",
    "assert DNA4.func_cg_fraction(sq1) == 5/24\n",
    "assert DNA4.func_cg_fraction(sq2) == 0/21\n",
    "assert DNA4.func_cg_fraction(sq3) == 0\n",
    "assert DNA4.func_cg_fraction(sq4) == 10/20\n",
    "\n",
    "# Test Cases 9-12: check annealing temperature\n",
    "assert DNA4.func_temperature(sq1) == 58\n",
    "assert DNA4.func_temperature(sq2) == 42\n",
    "assert DNA4.func_temperature(sq3) == 0\n",
    "assert DNA4.func_temperature(sq4) == 60\n",
    "\n",
    "# Test Cases 13-16: check runs\n",
    "assert DNA4.func_count_runs(sq1) == 2\n",
    "assert DNA4.func_count_runs(sq2) == 0\n",
    "assert DNA4.func_count_runs(sq3) == 0\n",
    "assert DNA4.func_count_runs(sq4) == 1\n",
    "\n",
    "# Test Cases 17-20: check repeats\n",
    "assert DNA4.func_count_repeats(sq1) == 0\n",
    "assert DNA4.func_count_repeats(sq2) == 12\n",
    "assert DNA4.func_count_repeats(sq3) == 0\n",
    "assert DNA4.func_count_repeats(sq4) == 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 6\n",
    "\n",
    "The following methods implement the cost functions for each individual criterion. Penalties will be applied to the properties that do not fall within the acceptable limits, according to the mathematical expressions given in the handout.\n",
    "\n",
    "The penalty parameter 𝛼 is an adjustable value to be decided. A larger penalty value penalizes deviations from the acceptable limits more. Every cost function will have its own penalty parameter."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "class PrimerDesign(PrimerDesign):\n",
    "    \n",
    "    '''Primer length is usually between 18 and 22 nucleotides'''\n",
    "    def cost_length(self, sq):\n",
    "        \n",
    "        # calculate the primer length\n",
    "        sq_len = len(sq)\n",
    "        \n",
    "        # penalization if primer length > max length\n",
    "        if(sq_len > self.max_length): \n",
    "            return (sq_len - self.max_length)*self.penalty_length\n",
    "        \n",
    "        # no penalization if min length <= primer length <= max length\n",
    "        elif(sq_len >= self.min_length): \n",
    "            return 0\n",
    "        \n",
    "        # penalization if primer length < min length\n",
    "        else:\n",
    "            return (self.min_length - sq_len)*self.penalty_length\n",
    "    \n",
    "    '''The primer annealing temperature is usually between 55°C and 63°C'''\n",
    "    def cost_temperature(self, sq):\n",
    "        \n",
    "        # calculate the primer annealing temp\n",
    "        temp = self.func_temperature(sq)\n",
    "        \n",
    "        # penalization if primer annealing temp > max temp\n",
    "        if(temp > self.max_temp):\n",
    "            return (temp - self.max_temp)*self.penalty_temp\n",
    "        \n",
    "        # no penalization if min temp <= primer annealing temp <= max temp\n",
    "        elif(temp >= self.min_temp):\n",
    "            return 0\n",
    "        \n",
    "        # penalization if primer annealing temp < min temp\n",
    "        else:\n",
    "            return (self.min_temp - temp)*self.penalty_temp \n",
    "    \n",
    "    '''The CG content is usually between 40% and 60%'''\n",
    "    def cost_cgcontent(self,sq):\n",
    "        \n",
    "        # calculate the primer % cg content\n",
    "        cg_content = self.func_cg_fraction(sq)\n",
    "        \n",
    "        # penalization if primer % cg content > max % cg content\n",
    "        if(cg_content > self.max_cg):\n",
    "            return (cg_content - self.max_cg)*self.penalty_cg\n",
    "        \n",
    "        # no penalization if min % cg content <= primer % cg content <= max % cg content\n",
    "        elif(cg_content >= self.min_cg):\n",
    "            return 0\n",
    "        \n",
    "        # penalization if % cg content < min % cg content\n",
    "        else:\n",
    "            return (self.min_cg - cg_content)*self.penalty_cg\n",
    "        \n",
    "    '''The difference in annealing temperatures between forward and reverse primers is usually less than 2°C'''\n",
    "    def cost_temperature_difference(self, fp, rp):\n",
    "        \n",
    "        # calculate the difference in the two primers' annealing temp\n",
    "        temp_diff = abs(self.func_temperature(fp) - self.func_temperature(rp)) # always a positive number\n",
    "        \n",
    "        # calculate the max allowable annealing temp diff\n",
    "        max_temp_diff = self.max_tdiff - self.min_tdiff\n",
    "        \n",
    "        # penalization if temp difference > max temp difference\n",
    "        if(temp_diff > max_temp_diff):\n",
    "            return (temp_diff - max_temp_diff)*self.penalty_tdiff\n",
    "        \n",
    "        # no penalization if temp difference <= max temp difference\n",
    "        else:\n",
    "            return 0\n",
    "    \n",
    "    '''The primer ought to be specific, i.e. only bind to one position on the dna sequence'''\n",
    "    def cost_specificity(self, sq):\n",
    "        \n",
    "        # calculate the no of repeats of primer sequence in the dna sequence\n",
    "        n_positions = self.func_count_specificity(sq)\n",
    "        \n",
    "        # penalization for every repeat\n",
    "        return (n_positions-1)*self.penalty_specificity\n",
    "        \n",
    "    '''Runs of the same nucleotide in the primer should be avoided'''\n",
    "    '''The maximum acceptable base run is usually 4'''\n",
    "    def cost_runs(self, sq):\n",
    "        \n",
    "        # calculate the no of runs that exceed the acceptable threshold\n",
    "        n_runs = self.func_count_runs(sq)\n",
    "        \n",
    "        # penalization for every run that exceeds the acceptable threshold\n",
    "        return n_runs*self.penalty_runs\n",
    "    \n",
    "    '''Repeating sequences of two bases in the primer should be avoided'''\n",
    "    def cost_repeats(self,sq):\n",
    "        max_repeats = self.repeat_threshold\n",
    "        \n",
    "        # calculate the no of repeated two-bases in the primer\n",
    "        n_repeats = self.func_count_repeats(sq)\n",
    "        \n",
    "        # penalization for every repeat that exceeds the acceptable threshold\n",
    "        if n_repeats > max_repeats:\n",
    "            return (n_repeats - max_repeats)*self.penalty_repeats\n",
    "        \n",
    "        # no penalization if n_repeats is within acceptable theshold\n",
    "        else:\n",
    "            return 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Testing methods in Task 6'''\n",
    "\n",
    "# primer sequences taken from eDimension test cases\n",
    "DNA6 = PrimerDesign(\"DNA6\")\n",
    "\n",
    "# note that DNA6.dna_sequence contains 2 fp sequences and 1 rp sequence\n",
    "DNA6.set_dna_sequence('atatgtatcgcattaaaaatacgaaaaaagtcatattcgcatcgcattaaaaatacgaaaaaagtcattatagcgatatatatattttatatataa')\n",
    "fp = 'cattaaaaatacgaaaaaagtcat'\n",
    "rp = 'atatatatattttatatataa'\n",
    "\n",
    "'''Primer length is usually between 18 and 22 nucleotides'''\n",
    "DNA6.max_length = 22\n",
    "DNA6.min_length = 18\n",
    "\n",
    "'''The difference in annealing temperatures between forward and reverse primers is usually less than 2°C'''\n",
    "DNA6.max_tdiff = 2\n",
    "DNA6.min_tdiff = 0\n",
    "\n",
    "'''The CG content is usually between 40% and 60%'''\n",
    "DNA6.max_cg = 0.6\n",
    "DNA6.min_cg = 0.4\n",
    "\n",
    "'''The primer annealing temperature is usually between 55°C and 63°C'''\n",
    "DNA6.max_temp = 63\n",
    "DNA6.min_temp = 55\n",
    "\n",
    "'''The maximum acceptable base run is usually 4'''\n",
    "DNA6.run_threshold = 4\n",
    "\n",
    "'''Repeating sequences of two bases in the primer should be avoided'''\n",
    "DNA6.repeat_threshold = 0\n",
    "\n",
    "# Test Cases 1-2: check length cost function\n",
    "assert DNA6.cost_length(fp) == 10*(24-22)\n",
    "assert DNA6.cost_length(rp) == 0\n",
    "\n",
    "# Test Case 3: check annealing temp cost function\n",
    "assert DNA6.cost_temperature_difference(fp,rp) == 10*(abs(58-42)-2)\n",
    "\n",
    "# Test Cases 4-5: check cgcontent cost function\n",
    "assert DNA6.cost_cgcontent(fp) == 10*(0.4-5/24)\n",
    "assert DNA6.cost_cgcontent(rp) == 10*(0.4-0)\n",
    "\n",
    "# Test Cases 6-7: check annealing temp cost function\n",
    "assert DNA6.cost_temperature(fp) == 0\n",
    "assert DNA6.cost_temperature(rp) == 10*(55-42)\n",
    "\n",
    "# Test Cases 8-9: check run cost function\n",
    "assert DNA6.cost_runs(fp) == 10*2\n",
    "assert DNA6.cost_runs(rp) == 0\n",
    "\n",
    "# Test Cases 10-11: check repeat cost function\n",
    "assert DNA6.cost_repeats(fp) == 0\n",
    "assert DNA6.cost_repeats(rp) == 10*(12-0)\n",
    "\n",
    "# Test Cases 12-13: check cost specificity\n",
    "assert DNA6.cost_specificity(fp) == 10*(2-1)\n",
    "assert DNA6.cost_specificity(rp) == 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 8\n",
    "\n",
    "After choosing a pair of primers, all its properties are input into an objective function. This objective function cost_objective_function is the sum of all the cost functions described in Task 5, which outputs a score that is indicative of the closeness of these primers to the optimal primer. A score of zero means that the primers selected have met all the criteria. A score that is close to zero means that the primers selected is close to the optimal choice."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "class PrimerDesign(PrimerDesign):\n",
    "    \n",
    "    def cost_objective_function(self, fp, rp):\n",
    "        '''complete the calculation of the cost'''\n",
    "        \n",
    "        cost = self.cost_length(fp) + self.cost_length(rp) + self.cost_temperature(fp) + self.cost_temperature(rp) + \\\n",
    "        self.cost_cgcontent(fp) + self.cost_cgcontent(rp) + self.cost_temperature_difference(fp,rp) + \\\n",
    "        self.cost_specificity(fp) + self.cost_specificity(rp) + self.cost_runs(fp) + self.cost_runs(rp) + \\\n",
    "        self.cost_repeats(fp) + self.cost_repeats(rp)\n",
    "        \n",
    "        return cost"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Testing method in Task 8'''\n",
    "\n",
    "# primer sequences taken from eDimension test cases (same as Task 6)\n",
    "# the output cost should be the sum of all the cost functions in Task 6\n",
    "\n",
    "DNA8 = PrimerDesign(\"DNA8\")\n",
    "\n",
    "# note that DNA8.dna_sequence contains 2 fp sequences and 1 rp sequence\n",
    "DNA8.set_dna_sequence('atatgtatcgcattaaaaatacgaaaaaagtcatattcgcatcgcattaaaaatacgaaaaaagtcattatagcgatatatatattttatatataa')\n",
    "fp = 'cattaaaaatacgaaaaaagtcat'\n",
    "rp = 'atatatatattttatatataa'\n",
    "\n",
    "'''Primer length is usually between 18 and 22 nucleotides'''\n",
    "DNA8.max_length = 22\n",
    "DNA8.min_length = 18\n",
    "\n",
    "'''The difference in annealing temperatures between forward and reverse primers is usually less than 2°C'''\n",
    "DNA8.max_tdiff = 2\n",
    "DNA8.min_tdiff = 0\n",
    "\n",
    "'''The CG content is usually between 40% and 60%'''\n",
    "DNA8.max_cg = 0.6\n",
    "DNA8.min_cg = 0.4\n",
    "\n",
    "'''The primer annealing temperature is usually between 55°C and 63°C'''\n",
    "DNA8.max_temp = 63\n",
    "DNA8.min_temp = 55\n",
    "\n",
    "'''The maximum acceptable base run is usually 4'''\n",
    "DNA8.run_threshold = 4\n",
    "\n",
    "'''Repeating sequences of two bases in the primer should be avoided'''\n",
    "DNA8.repeat_threshold = 0\n",
    "\n",
    "assert DNA8.cost_objective_function(fp,rp) == 10*(24-22) + 0 + 10*(abs(58-42)-2) + 10*(0.4-5/24) + 10*(0.4-0) + 0 + 10*(55-42) + 10*2 + 0 + 0 + 10*(12-0) + 10*(2-1) + 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 9\n",
    "\n",
    "The method cost_objective_function_info provides information on which criteria was met or not met. If the cost function for a particular criterion returns zero, the criterion is met. Note that in this output, the reverse primer displayed is the reverse complement of what is selected from the DNA sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def checkCriterion(cost):\n",
    "    if cost == 0:\n",
    "        return \"MET\"\n",
    "    else:\n",
    "        return \"NOT MET\"\n",
    "\n",
    "def ReverseComplement(rp):    \n",
    "        \n",
    "        # so far we have working with rp selected directly from the DNA sequence given\n",
    "        # hence we need to reverse complement the input rp to output the correct rp\n",
    "        new_rp = \"\"\n",
    "        for base in rp:\n",
    "            if base == 'a':\n",
    "                new_rp += 't'\n",
    "            elif base == 't':\n",
    "                new_rp += 'a'\n",
    "            elif base == 'c':\n",
    "                new_rp += 'g'\n",
    "            elif base == 'g':\n",
    "                new_rp += 'c'\n",
    "        \n",
    "        return new_rp[::-1]\n",
    "\n",
    "class PrimerDesign(PrimerDesign):\n",
    "    \n",
    "    def cost_objective_function_info(self, fp, rp):\n",
    "        \n",
    "        # obtain the correct rp\n",
    "        correct_rp = ReverseComplement(rp)\n",
    "        \n",
    "        # forward and reverse primer info\n",
    "        # note that for the rp, the cost function will still remain unchanged whether it is reverse complemented\n",
    "        for i in [fp,rp]:\n",
    "            if i == fp:\n",
    "                print(\"===Forward Primer=== {}\".format(fp))\n",
    "            else:\n",
    "                print(\"===Reverse Primer=== {}\".format(correct_rp))\n",
    "            print(\"Criterion{:>26}{:>25}{:>18}\".format(\"Data\", \"Cost function score\", \"MET/ NOT MET\"))\n",
    "            print(\"------------------------------------------------------------------------------\")\n",
    "            print(\"length{:>29.3f}{:>25.3f}{:>18}\".format(self.func_length(i), self.cost_length(i), checkCriterion(self.cost_length(i))))\n",
    "            print(\"annealing temperature{:>14.3f}{:>25.3f}{:>18}\".format(self.func_temperature(i), self.cost_temperature(i), checkCriterion(self.cost_temperature(i))))\n",
    "            print(\"%cg_content{:>24.3f}{:>25.3f}{:>18}\".format(self.func_cg_fraction(i), self.cost_cgcontent(i), checkCriterion(self.cost_cgcontent(i))))\n",
    "            print(\"specificity{:>24.3f}{:>25.3f}{:>18}\".format(self.func_count_specificity(i), self.cost_specificity(i), checkCriterion(self.cost_specificity(i))))\n",
    "            print(\"runs{:>31.3f}{:>25.3f}{:>18}\".format(self.func_count_runs(i), self.cost_runs(i), checkCriterion(self.cost_runs(i))))\n",
    "            print(\"repeats{:>28.3f}{:>25.3f}{:>18}\\n\".format(self.func_count_repeats(i), self.cost_repeats(i), checkCriterion(self.cost_repeats(i))))\n",
    "        \n",
    "        # temperature difference between forward and reverse primers\n",
    "        temp_diff = abs(self.func_temperature(fp) - self.func_temperature(rp))\n",
    "        print(\"Temperature Difference{:>13.3f}{:>25.3f}{:>18}\\n\".format(temp_diff, self.cost_temperature_difference(fp,rp), checkCriterion(self.cost_temperature_difference(fp,rp))))\n",
    "        \n",
    "        # cost objective function\n",
    "        cost_function = self.cost_objective_function(fp,rp)\n",
    "        print(\"The calculated cost objective function value is {:.3f}\".format(self.cost_objective_function(fp,rp)))\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part 2 - Implement the Simulated Annealing Algorithm"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 12\n",
    "\n",
    "Implement the simulated annealing algorithm to search for the primers that give the best possible score."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "class PrimerDesign(PrimerDesign): \n",
    "\n",
    "    def func_simulated_annealing(self):\n",
    "        \n",
    "        # step 1.1: deciding simulation parameters\n",
    "        temperature = self.initial_temperature\n",
    "        stopping_temperature = self.stopping_temperature\n",
    "        drop = self.drop_fraction\n",
    "        \n",
    "        # step 1.2: choosing a possible solution\n",
    "        fp_current = self.func_select_random(sqtype='forward')\n",
    "        rp_current = self.func_select_random(sqtype='reverse')\n",
    "        current_cost = self.cost_objective_function(fp_current, rp_current)\n",
    "        \n",
    "        iteration = 0\n",
    "        while(temperature > stopping_temperature):\n",
    "            # step 2: obtain a new possible solution\n",
    "            fp_new = self.get_primer_neighbour(fp_current)\n",
    "            rp_new = self.get_primer_neighbour(rp_current)\n",
    "            new_cost = self.cost_objective_function(fp_new, rp_new)\n",
    "            \n",
    "            # step 3: accept new soln if it is a lower cost than the current soln\n",
    "            if(new_cost < current_cost):\n",
    "                fp_current = fp_new\n",
    "                rp_current = rp_new\n",
    "                current_cost = new_cost\n",
    "            \n",
    "            # step 4: if the current soln is worse, we may still accept it\n",
    "            else:\n",
    "                # step 4a: calculate an acceptance probability\n",
    "                delta = new_cost - current_cost\n",
    "                acceptance_probability = math.exp(-delta/temperature)\n",
    "                \n",
    "                # step 4b: throw a dice\n",
    "                num = random.uniform(0,1)\n",
    "                \n",
    "                # step 4c: accept the new soln if num < acceptance probability\n",
    "                if(acceptance_probability > num):\n",
    "                    fp_current = fp_new\n",
    "                    rp_current = rp_new\n",
    "                    current_cost = new_cost\n",
    "                    \n",
    "                # step 5: decrease the temp by a small factor\n",
    "                temperature = temperature*drop\n",
    "            iteration += 1\n",
    "        \n",
    "        self.cost_objective_function_info(fp_current, rp_current)\n",
    "        print(\"The number of iterations required is {}.\".format(iteration))\n",
    "        \n",
    "        # check primer locations\n",
    "        if fp_current not in self.dna_sequence or rp_current not in self.dna_sequence:\n",
    "            return \"Error\" # primer not found\n",
    "        else:\n",
    "            print(\"The forward primer location is from {} to {}.\".format(self.dna_sequence.index(fp_current), self.dna_sequence.index(fp_current) + len(fp_current)))\n",
    "            print(\"The reverse primer location is from {} to {}.\".format(self.dna_sequence.index(rp_current), self.dna_sequence.index(rp_current) + len(rp_current)))\n",
    "        \n",
    "        # expected PCR length\n",
    "        print(\"The expcted PCR length should be {} nucleotides long.\".format(self.dna_sequence.index(rp_current) + len(rp_current) - self.dna_sequence.index(fp_current) + 1))\n",
    "        \n",
    "    def get_primer_neighbour(self, sq):\n",
    "        new_start = self.dna_sequence.index(sq) + random.randint(-10,10)\n",
    "        new_primer_length = random.randint(self.min_length, self.max_length)\n",
    "        new_primer = self.dna_sequence[new_start: new_start + new_primer_length + 1]\n",
    "        return new_primer\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Store the DNA sequence given to you in the variable below "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "dna_sequence = '''1 gattggctgg cgcggattcc agctgctttc caagtcagcg gcgcctagtg agagtcaggg\n",
    "61 gggcccggcc cgcgccctcc ccgcccagcc gcctccccgt cgacgcccag ggctggggcg\n",
    "121 agccaggctg cctttcgaac ttggggggct tctcctcttg tctcccactg gtgctctggc\n",
    "181 tgtgaatcca tccaggggcc aggatgacaa tccgacacca aggccagcag tacaggccga\n",
    "241 ggatggcatt tctccaaaag attgaagcgc tcgtgaagga catgcagaac ccagagacag\n",
    "301 gggtccgaat gcagaaccag agggtcctgg tcaccagcgt tcctcatgcc atgacaggaa\n",
    "361 gtgatgttct gcaatggatc gtccagcggc tttggatctc cagtctggag gcacagaact\n",
    "421 tgggcaactt tattgtcagg tatggctaca tttaccccct gcaagacccc aagaatctca\n",
    "481 ttctcaagcc tgatggcagc ctctacagat ttcagacacc gtatttctgg cccacccagc\n",
    "541 agtggccagc tgaagatacc gattacgcca tctatctggc caagcgaaat atcaaaaaga\n",
    "601 aagggatttt ggaagaatat gaaaaggaaa attacaattt cttgaaccaa aaaatgaact\n",
    "661 ataagtggga ctttgtcatt atgcaggcca aagagcagta cagggctgga aaggagagga\n",
    "721 acaaagcaga cagatatgcc ctggactgcc aggagaaggc atactggctg gtgcaccgat\n",
    "781 gccctcctgg aatggacaat gtgctggact acggcctgga ccgagtgacc aatccgaatg\n",
    "841 aagtcaaggt aaaccagaaa caaacagtcg ttgctgtcaa aaaagagatc atgtattacc\n",
    "901 aacaggcctt gatgaggtcc acagtgaagt cttctgtgtc cctgggaggg attgtgaaat\n",
    "961 acagtgagca gttctcatcc aacgatgcca tcatgtcagg ctgcctcccc agcaacccct\n",
    "1021 ggatcaccga tgacacccag ttctgggact taaatgccaa attggtggaa atcccaacca\n",
    "1081 agatgcgagt ggaacgatgg gccttcaact tcagcgaatt gatccgagac cccaaaggtc\n",
    "1141 gacagagctt ccagtacttc ctcaagaaag ggctcagtgg agagaatctg ggattctggg\n",
    "1201 aagcctgcga ggatctgaag tatggagatc agtccaaagt caaggagaaa gcagaggaga\n",
    "1261 tttacaagct gttcctggcc ccgggggcga ggcgctggat caacatagat ggcaaaacca\n",
    "1321 tggacatcac agtgaagggg ctgaagcacc cccaccgcta tgtgctggac gccgcacaaa\n",
    "1381 cccacattta catgctcatg aagaaggatt cttatgctcg ctatttaaaa tctccgatct\n",
    "1441 ataaggacat gctggccaaa gctattgaac ctcaggaaac caccaagaaa agctccaccc\n",
    "1501 tcccttttat gcggcgtcac ctgcgctcca gcccaagccc tgtcatcctg agacagctgg\n",
    "1561 aagaggaagc caaggcccga gaagcagcca acactgtgga catcacccag ccgggccagc\n",
    "1621 acatggctcc cagcccccat ctgaccgtgt acaccgggac ctggaggccc ccgtctcctt\n",
    "1681 ctagcccctt ctcctcctcc tgccgctccc ccaggaagcc tttcgcctca cccagccgct\n",
    "1741 tcatccggcg acccagcacc accatctgcc cctcacccat cagagtggcc ttggagagct\n",
    "1801 catcgggctt ggagcagaaa ggggagtgca gcgggtccat ggccccccgt gggccctctg\n",
    "1861 tcaccgagag cagcgaggcc tccctcgaca cctcctggcc tcgcagccgg cccagggccc\n",
    "1921 ctcctaaggc ccgcatggct ctgtccttca gcaggtttct gagacgaggc tgtctggcct\n",
    "1981 cacctgtctt tgccaggctc tcacccaagt gccctgctgt gtcccacggg agggtgcagc\n",
    "2041 ccctggggga cgtgggccag cagctgccac gattgaaatc caagagagta gcaaactttt\n",
    "2101 tccagatcaa aatggatgtg cccacgggga gcgggacctg cttgatggac tcggaggatg\n",
    "2161 ctggaacagg agagtcgggt gaccgggcca cagaaaagga ggtcatctgc ccctgggaga\n",
    "2221 gcctgtaagg aaagaggcag gctgagctgg gggctctgga ccaggaagat gctctgacag\n",
    "2281 atgccatggt atgggccaca ggacacactt gctcgagaac caaagtgcat ttgggtgaca\n",
    "2341 tttgaagatt ggggagacaa gatggggtag attgtggcaa agaatgctct ggctggttac\n",
    "2401 caggggccaa ctccttctcc tcttcctgac cctccctccc ctgggcagaa gaaacgcatg\n",
    "2461 tggaccagaa gactttccct gctgccttaa aaccaataaa aggttaactt taagtttctt\n",
    "2521 ggaaaaaaaa aaaaaaaaaa\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Instantiate your class and read in the DNA sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "DNA_test = PrimerDesign(\"DNA_test\")\n",
    "DNA_test.set_dna_sequence(dna_sequence)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### If you need to adjust any parameter from their default values in the init method, do it here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''parameters for the length criterion'''\n",
    "'''Primer length is usually between 18 and 22 nucleotides'''\n",
    "# This length is long enough for adequate specificity\n",
    "# and short enough for primers to bind easily to the template at the annealing temperature.\n",
    "DNA_test.max_length = 22\n",
    "DNA_test.min_length = 18\n",
    "DNA_test.penalty_length = 10\n",
    "\n",
    "'''parameters for the temperature difference criterion'''\n",
    "'''The difference in annealing temperatures between forward and reverse primers is usually less than 2°C.'''\n",
    "# This is because during the PCR annealing step, there is only one annealing temperature where both the reverse and\n",
    "# forward primers are expected to anneal to the denatured single DNA strand, so the difference cannot be too far apart\n",
    "# if not the DNA amplification by Taq polymerase during the elongation step cannot happen\n",
    "DNA_test.max_tdiff = 2\n",
    "DNA_test.min_tdiff = 0\n",
    "# higher weightage because of the reason described above\n",
    "DNA_test.penalty_tdiff = 30\n",
    "\n",
    "'''parameters for the cg content criterion'''\n",
    "'''The CG content is usually between 40% and 60%'''\n",
    "DNA_test.max_cg = 0.6\n",
    "DNA_test.min_cg = 0.4\n",
    "# higher weightage so that the order of the magnitude of the score is comparable to the rest\n",
    "DNA_test.penalty_cg = 100 \n",
    "\n",
    "'''parameters for the annealing temperature criterion'''\n",
    "'''The primer annealing temperature is usually between 55°C and 63°C'''\n",
    "DNA_test.max_temp = 63\n",
    "DNA_test.min_temp = 55\n",
    "# less weightage because the annealing temp is calculated using the cg_fraction\n",
    "# so the penalty would have been imposed on the cg_content cost function\n",
    "DNA_test.penalty_temp = 10\n",
    "\n",
    "'''parameters for the run criterion'''\n",
    "'''The maximum acceptable base run is usually 4'''\n",
    "DNA_test.run_threshold = 4\n",
    "# less weightage because we assume the other sequences\n",
    "# to be sufficiently different for the primer to recognise and bind\n",
    "DNA_test.penalty_runs = 10\n",
    "\n",
    "'''parameters for the repeat criterion'''\n",
    "'''Repeating sequences of two bases in the primer should be avoided'''\n",
    "DNA_test.repeat_threshold = 0\n",
    "# less weightage because we have already made the requirement stricter\n",
    "# by setting the threshold to be as low as possible i.e. 0\n",
    "DNA_test.penalty_repeats = 10\n",
    "\n",
    "'''parameters for the specificity criterion'''\n",
    "# higher weightage because we want the chosen primer to bind at a specific location\n",
    "# if not the PCR product size will be different and RFLP gel electrophoresis will be hard\n",
    "DNA_test.penalty_specificity = 20 \n",
    "\n",
    "# our dna sequence shows a mutated sequence at the 1100th position\n",
    "# so we design our forward and reverse primers should be at least 300bp before and after the identified region\n",
    "# so that the final PCR product is around 300 - 1000 bp\n",
    "# gene sequencing and identification can then be performed on this length\n",
    "# this is because 100 codons are needed as the shortest length of a putative gene, so at least 300bp\n",
    "# also, the PCR length should be sufficiently long to be run and easily identified on the gel\n",
    "# after performing RFLP enzyme digestion and gel electrophoresis\n",
    "\n",
    "'''locations where the forward primer should be chosen from'''\n",
    "DNA_test.fp_start = 600\n",
    "DNA_test.fp_end = 700\n",
    "\n",
    "'''locations where the reverse primer should be chosen from'''\n",
    "DNA_test.rp_start = 1600\n",
    "DNA_test.rp_end = 1700"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Show the outcome of your testing and the functions in the subsequent cells "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "===Forward Primer=== aatttcttgaaccaaaaaat\n",
      "Criterion                      Data      Cost function score      MET/ NOT MET\n",
      "------------------------------------------------------------------------------\n",
      "length                       20.000                    0.000               MET\n",
      "annealing temperature        48.000                   70.000           NOT MET\n",
      "%cg_content                   0.200                   20.000           NOT MET\n",
      "specificity                   1.000                    0.000               MET\n",
      "runs                          1.000                   10.000           NOT MET\n",
      "repeats                       0.000                    0.000               MET\n",
      "\n",
      "===Reverse Primer=== ctgggagccatgtgctggcc\n",
      "Criterion                      Data      Cost function score      MET/ NOT MET\n",
      "------------------------------------------------------------------------------\n",
      "length                       20.000                    0.000               MET\n",
      "annealing temperature        68.000                   50.000           NOT MET\n",
      "%cg_content                   0.700                   10.000           NOT MET\n",
      "specificity                   1.000                    0.000               MET\n",
      "runs                          0.000                    0.000               MET\n",
      "repeats                       1.000                   10.000           NOT MET\n",
      "\n",
      "Temperature Difference       20.000                  540.000           NOT MET\n",
      "\n",
      "The calculated cost objective function value is 710.000\n"
     ]
    }
   ],
   "source": [
    "'''Testing method for Task 9'''\n",
    "\n",
    "fp = DNA_test.func_select_random()\n",
    "rp = DNA_test.func_select_random(sqtype='reverse')\n",
    "DNA_test.cost_objective_function_info(fp,rp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "===Forward Primer=== gagcagtacagggctggaaa\n",
      "Criterion                      Data      Cost function score      MET/ NOT MET\n",
      "------------------------------------------------------------------------------\n",
      "length                       20.000                    0.000               MET\n",
      "annealing temperature        62.000                    0.000               MET\n",
      "%cg_content                   0.550                    0.000               MET\n",
      "specificity                   1.000                    0.000               MET\n",
      "runs                          0.000                    0.000               MET\n",
      "repeats                       0.000                    0.000               MET\n",
      "\n",
      "===Reverse Primer=== aggagaaggggctagaagg\n",
      "Criterion                      Data      Cost function score      MET/ NOT MET\n",
      "------------------------------------------------------------------------------\n",
      "length                       19.000                    0.000               MET\n",
      "annealing temperature        60.000                    0.000               MET\n",
      "%cg_content                   0.579                    0.000               MET\n",
      "specificity                   1.000                    0.000               MET\n",
      "runs                          0.000                    0.000               MET\n",
      "repeats                       1.000                   10.000           NOT MET\n",
      "\n",
      "Temperature Difference        2.000                    0.000               MET\n",
      "\n",
      "The calculated cost objective function value is 10.000\n",
      "The number of iterations required is 10325.\n",
      "The forward primer location is from 692 to 712.\n",
      "The reverse primer location is from 1676 to 1695.\n",
      "The expcted PCR length should be 1004 nucleotides long.\n"
     ]
    }
   ],
   "source": [
    "'''Testing method for Task 10'''\n",
    "\n",
    "DNA_test.func_simulated_annealing()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What were the simulation parameters that you used (in Step 1.1 of the algorithm)?\n",
    "\n",
    "self.initial_temperature = 200\n",
    "\n",
    "self.stopping_temperature = 0.01\n",
    "\n",
    "self.drop_fraction = 0.999\n",
    "        \n",
    "### From the output of your code, how well does the final solution meet all the criteria?\n",
    "\n",
    "The final solution meets all the criteria very well as the individual cost function score is 0.000 for all the criterion.\n",
    "\n",
    "### How many iterations were required to obtain this solution?\n",
    "\n",
    "10325 iterations were required to obtain this solution.\n",
    "\n",
    "### Will a brute-force algorithm (one that considers every single possible solution) require fewer or more iterations? Explain.\n",
    "\n",
    "A brute-force algorithm will require more iterations. As a rough calculation for the brute force algorithm using the 2.5kbp long DNA sequence provided, assuming that we begin by displacing the forward and reverse primers from the halfway point, conservatively we need at least 1000^1000 iterations, which is much more than the simulated annealing algorithm.\n",
    "\n",
    "The simulated annealing algorithm checks the cost function score against certain criteria, so is able to very quickly approach the optimal solution."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
